### DISCUSSION TITLE
[Bug] SkyWalking Java agent plugin InstanceMethodsInterceptPoint method match failed

### DISCUSSION THREAD

**Discussion by creeew** [apache/skywalking#8562] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

I following the doc to build my SkyWalking plugin in Java agent, unfortunately there are some ElementMatchers match way cannot work correctly.
In InstanceMethodsInterceptPoint#getMethodsMatcher everything is alright if you using ElementMatchers.named to match a method. But in nameContains/nameEndsWith/nameMatches cannot match the method which should be match.

### What you expected to happen

InstanceMethodsInterceptPoint match method correctlly

### How to reproduce

`
public class TestInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {

    private static final String ENHANCE_CLASS = "test.plugin.TestService";
    private static final String INTERCEPT_CLASS = "org.apache.skywalking.apm.plugin.test.TestInterceptor";

    @Override
    protected ClassMatch enhanceClass() {
        return NameMatch.byName(ENHANCE_CLASS);
    }

    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        return null;
    }

    @Override
    public InstanceMethodsInterceptPoint[] getInstanceMethodsInterceptPoints() {
        return new InstanceMethodsInterceptPoint[] {
            new InstanceMethodsInterceptPoint() {
                @Override
                public ElementMatcher<MethodDescription> getMethodsMatcher() {
                    // success
                    //  return named("testPublicMethod");
                    // return ElementMatchers.nameStartsWith("t");
                    // return ElementMatchers.nameMatches("testPublicMethod");

                    // fail
                    return ElementMatchers.nameContains("t");
                   // return ElementMatchers.nameEndsWith("d");
                   // return ElementMatchers.nameMatches("[a-zA-Z0-9_$]+");
                }

                @Override
                public String getMethodsInterceptor() {
                    return INTERCEPT_CLASS;
                }

                @Override
                public boolean isOverrideArgs() {
                    return false;
                }
            }
        };
    }
}
`
my method name is "testPublicMethod", those failed code like nameContain and endsWith should be match my method.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Is that method a class method or static method? Is there a plugin working on that method already?(such as Spring annotation plugin) Have you checked whether thid instrumentation activated? Logs or debug?

It is hard to tell, you provided very limited context.

**Reply by wu-sheng** [reply]
Please don't submit an issue, as hundreds of plugins are running to intercept methods. If this simple way had a bug, most things are failing.

**Reply by creeew** [reply]
It's a class instance method. 
A lot of plugins using ElementMatchers.named and this working and also doing right in my test code. But ElementMatchers.nameContains cannot work correctly in my test scenario. I've tested twice after you replied.

**Reply by wu-sheng** [reply]
If you are asking about `ElementMatchers#nameContains` only, then maybe you asked from the wrong place? This is a bytebuddy native method, not implemented by SkyWalking Java agent. 

Also in our agent codes, we only use this matcher for class name once at here 
https://github.com/apache/skywalking-java/blob/f0c22f32dd1493c03b01013ddac93e861902d153/apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java#L95
So, can't tell much about your scenario.

**Reply by creeew** [reply]
Not just a nameContains and nameEndsWith cannot work correctly. And I've tested in byteBuddy way to enhance my method, the thing is native byteBuddy way worked but skywalking wrapped way failed in nameContains and nameEndsWith. Maybe you change a working plugin  match method by endsWith you will know the problem is.

**Reply by wu-sheng** [reply]
We never used `nameEndsWith`. But as no upstream works are working on this, we don't have experience with this. Personally, I don't use Java agent, maybe someone is willing to try. I just shared the current status with you.

**Reply by wu-sheng** [reply]
All you defined, we are using like this, https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/interceptor/enhance/ClassEnhancePluginDefine.java#L132.
As you have already used ByteBuddy, you could find the difference. We have some conditions set in the kernel.
